clone is broken

clone does not work, as can be seen in this example:

```java
// clone.bsh
public class A implements Cloneable {
    String name;
    void whoami() {
        System.out.println("I am " + this + " my name is " + name);
    }
}

A a = new A();
a.name = "a";
a.whoami();

A b = a.clone();
b.name = "b";
b.whoami();
// end of clone.bsh
```
The output is:
```java
% java -cp bsh.jar bsh.Interpreter clone.bsh
I am A@589e56 my name is a
I am A@589e56 my name is a
```
This is with 2.0b4 and JVM 1.5.

\- Godmar


Reported by: gback

Original Ticket: "beanshell/bugs/283":https://sourceforge.net/p/beanshell/bugs/283

There does appear to be a problem here, although perhaps not the original issue.  When sourced the test script yields the following output:

```text
I am A@2e4553 my name is a
// Error: Target Exception: Method Invocation this.interpreter.source : at Line: 14 : in file: /bsh/commands/source.bsh : this .interpreter .source ( filename , this .caller .namespace )

Called from method: source : at Line: 1 : in file: <unknown file> : source ( "clone.bsh" ) Caused by: bsh.EvalError: Sourced file: clone.bsh : Typed variable declaration : Error in method invocation: Method clone() not found in class'A' : at Line: 13 : in file: clone.bsh : a .clone ( )

bsh.ReflectError: Method clone() not found in class'A'
```

--- a/src/conf/version.properties
+++ b/src/conf/version.properties
@@ -14,6 +14,6 @@
 #
 #Project version number - populated by maven
 #Project build number - incremented by beanshell-maven-plugin
-#Wed Dec 21 23:25:32 SAST 2022
-build=1677
+#Fri Dec 23 02:45:54 SAST 2022
+build=1852
 release=${project.version}

--- a/src/main/java/bsh/BshMethod.java
+++ b/src/main/java/bsh/BshMethod.java
@@ -24,7 +24,6 @@
  *                                                                           *
  *****************************************************************************/
 
-
 package bsh;
 
 import java.util.stream.IntStream;
@@ -47,7 +46,7 @@ import java.lang.reflect.InvocationTargetException;
     Note: this method incorrectly caches the method structure.  It needs to
     be cleared when the classloader changes.
 */
-public class BshMethod implements Serializable {
+public class BshMethod implements Serializable, Cloneable {
 
     private static final long serialVersionUID = 1L;
 
@@ -120,6 +119,17 @@ public class BshMethod implements Serializable {
 
         this.javaMethod = method;
         this.javaObject = object;
+    }
+
+    /** Cloneable clone method implementation, delegated to parent.
+     * {@inheritDoc} */
+    @Override
+    public BshMethod clone() {
+        BshMethod returnClone = null;
+        try {
+            returnClone = (BshMethod) super.clone();
+        } catch (CloneNotSupportedException e) { /* is cloneable */ }
+        return returnClone;
     }
 
     /**
@@ -497,6 +507,14 @@ public class BshMethod implements Serializable {
                     "Incorrect type returned from method: "
                     + name + e.getMessage(), node, callstack );
             }
+        }
+
+        // when cloning a generated class deep copy This reference #421
+        if ("clone".equals(getName())) {
+            String className = ret.getClass().getSimpleName();
+            This thiz = Reflect.getClassInstanceThis(ret, className);
+            if (null != thiz) // not a generated class instance
+                return thiz.cloneMethodImpl(callerInfo, callstack, ret);
         }
 
         return ret;

--- a/src/main/java/bsh/NameSpace.java
+++ b/src/main/java/bsh/NameSpace.java
@@ -133,14 +133,12 @@ public class NameSpace
     Object getClassInstance() throws UtilEvalError {
         if (this.classInstance != null)
             return this.classInstance;
-        if (this.classStatic != null
-        // || (getParent()!=null && getParent().classStatic != null)
-       )
+        if (this.classStatic != null)
             throw new UtilEvalError(
-                    "Can't refer to class instance from static context.");
+                "Can't refer to class instance from static context.");
         else
             throw new InterpreterError(
-                    "Can't resolve class instance 'this' in: " + this);
+                "Can't resolve class instance 'this' in: " + this);
     }
     // End instance data
 
@@ -452,6 +450,14 @@ public class NameSpace
     */
     public String [] getVariableNames() {
         return this.variables.keySet().stream().toArray(String[]::new);
+    }
+
+    /**
+        Get the variables defined in this namespace.
+        (This does not show variables in parent namespaces).
+    */
+    public Variable [] getVariables() {
+        return this.variables.values().stream().toArray(Variable[]::new);
     }
 
     /**

--- a/src/main/java/bsh/This.java
+++ b/src/main/java/bsh/This.java
@@ -412,21 +412,9 @@ public final class This implements java.io.Serializable, Runnable
             return this == obj ? Boolean.TRUE : Boolean.FALSE;
         }
 
-        // a default clone()
+        // a default clone() method
         if ( methodName.equals("clone") && args.length==0 ) {
-            NameSpace ns = new NameSpace(namespace, namespace.getName()+" clone");
-            try {
-                for( String varName : namespace.getVariableNames() ) {
-                    ns.setLocalVariable(varName,
-                            namespace.getVariable(varName, false), false);
-                }
-                for( BshMethod method : namespace.getMethods() ) {
-                    ns.setMethod(method);
-                }
-            } catch ( UtilEvalError e ) {
-                throw e.toEvalError( Node.JAVACODE, callstack );
-            }
-            return ns.getThis(declaringInterpreter);
+            return cloneMethodImpl(callerInfo, callstack);
         }
 
         // Look for a default invoke() handler method in the namespace
@@ -446,6 +434,77 @@ public final class This implements java.io.Serializable, Runnable
                     " not found in bsh scripted object: "+ namespace.getName(),
                     callerInfo, callstack, e );
         }
+    }
+
+    /** Clone method implementation overload called with no cloned instance.
+     * @param callerInfo for error context
+     * @param callstack for error context
+     * @return returns a cloned instances with deep copied This reference
+     * @throws EvalError rolled up exceptions to user */
+    Object cloneMethodImpl(Node callerInfo, CallStack callstack) throws EvalError {
+        return cloneMethodImpl(callerInfo, callstack, null);
+    }
+
+    /** Clone method implementation called with cloned class instance. This
+     * method was extracted from invokeMethod to facilitate its reuse when
+     * generated classes implement Cloneable. The standard implementation
+     * succesfully produces a new instance object but it also copies This
+     * reference rendering two unique instances with the same context.#421
+     * A deep copy of everything is required to produce two independent
+     * instances, as would be expected, after a clone.
+     * Will also attempt to produce a clone if called from This context.
+     * @param callerInfo for error context
+     * @param callstack for error context
+     * @param clonedInstance a cloned instance or null to clone from This
+     * @return returns a cloned instances with deep copied This reference
+     * @throws EvalError rolled up exceptions to user */
+    Object cloneMethodImpl(Node callerInfo, CallStack callstack, Object clonedInstance) throws EvalError {
+        NameSpace ns = new NameSpace(namespace.getParent(), namespace.getName()+" clone");
+        try {
+            if (null == clonedInstance) { // no clone instance, attempt manual clone
+                if (null == namespace.classStatic) // thats not going to work, lets
+                    return ns.getThis(declaringInterpreter); // just return This.
+                // instantiate "clone" available class with empty constructor
+                clonedInstance = namespace.classStatic.getConstructor(new Class<?>[0])
+                    .newInstance(new Object[0]);
+            }
+            ns.setClassInstance(clonedInstance);
+            ns.setClassStatic(namespace.classStatic);
+            ns.isClass = true;
+            ns.isMethod = true;
+
+            for (Variable var : namespace.getVariables()) {
+                ns.setVariableImpl(var);
+            }
+
+            NameSpace declaringNS = new NameSpace(ns, null);
+            declaringNS.setClassInstance(clonedInstance);
+            declaringNS.setClassStatic(ns.classStatic);
+            declaringNS.isClass = true;
+            declaringNS.isMethod = true;
+
+            for (BshMethod method : namespace.getMethods()) {
+                BshMethod cloneM = method.clone();
+                cloneM.declaringNameSpace = declaringNS;
+                ns.setMethod(cloneM);
+            }
+
+            // update This reference in cloned instance
+            Reflect.getLHSObjectField(clonedInstance,
+                BSHTHIS + ns.classStatic.getSimpleName())
+                    .assign(ns.getThis(declaringInterpreter));
+
+        } catch (NoSuchMethodException | SecurityException | InstantiationException
+                | IllegalAccessException | IllegalArgumentException
+                | InvocationTargetException e) {
+            throw new EvalError("Unable to clone from This reference: "
+                + e.getMessage(), callerInfo, callstack);
+        } catch (UtilEvalError e) {
+            throw e.toEvalError("Unable to assign clone instance This: "
+                + e.getMessage(), callerInfo, callstack);
+        }
+
+        return clonedInstance;
     }
 
     /** Delegate method to return enum values array.

--- a/src/test/resources/test-scripts/cloneable.bsh
+++ b/src/test/resources/test-scripts/cloneable.bsh
@@ -1,0 +1,51 @@
+#!/bin/java bsh.Interpreter
+
+source("TestHarness.bsh");
+source("Assert.bsh");
+
+public class CanClone implements Cloneable {
+   String name = "original";
+
+   public Object clone() throws CloneNotSupportedException {
+       return super.clone();
+   }
+
+   String whoami() {
+       return "I am " + this + " my name is " + name;
+   }
+
+   Object getThis() {
+      return this;
+   }
+}
+
+cclne = new CanClone();
+clne = cclne.clone();
+
+assertNotEquals("Clones are not the same", cclne, clne);
+assertEquals("Clones are of the same type", cclne.getClass(), clne.getClass());
+
+assertNotEquals("Clones have different This", cclne._bshThisCanClone, clne._bshThisCanClone);
+assertNotEquals("Clones have different This namespace", cclne._bshThisCanClone.namespace, clne._bshThisCanClone.namespace);
+assertNotEquals("Clones have different whoami", cclne.whoami(), clne.whoami());
+assertNotEquals("Clones have different this", cclne.getThis(), clne.getThis());
+
+assertEquals("Clone's instance equals this", clne, clne.getThis());
+assertNotEquals("Clone's instance not equal to other clone's this", clne, cclne.getThis());
+assertThat("Clone's whoami contains instance string", clne.whoami(), containsString(clne.toString()));
+assertThat("Clone's whoami ends with name string", clne.whoami(), endsWith(clne.name));
+
+assertEquals("Clones have the same field values", cclne.name, clne.name);
+clne.name = "iClone";
+assertNotEquals("Clones can change fields independently", cclne.name, clne.name);
+assertEquals("Name is what we expect", "iClone", clne.name);
+assertThat("Name in method output", clne.whoami(), endsWith(clne.name));
+assertThat("Name still in other method output", cclne.whoami(), endsWith(cclne.name));
+cclne.name = "oClone";
+assertNotEquals("Clones can change fields independently", clne.name, cclne.name);
+assertEquals("Name is what we expect", "oClone", cclne.name);
+assertThat("Name in method output", cclne.whoami(), endsWith(cclne.name));
+assertThat("Name still in other method output", clne.whoami(), endsWith(clne.name));
+
+
+complete();

--- a/src/test/resources/test-scripts/clonethis.bsh
+++ b/src/test/resources/test-scripts/clonethis.bsh
@@ -1,0 +1,56 @@
+#!/bin/java bsh.Interpreter
+
+source("TestHarness.bsh");
+source("Assert.bsh");
+
+
+public class CannotClone {
+    String name = "original";
+
+    String whoami() {
+        return "I am " + this + " my name is " + name;
+    }
+
+    Object getThis() {
+       return this;
+    }
+ }
+
+ cclne = new CannotClone();
+ clne = cclne._bshThisCannotClone.clone();
+
+ assertNotEquals("Clones are not the same", cclne, clne);
+ assertEquals("Clones are of the same type", cclne.getClass(), clne.getClass());
+
+ assertNotEquals("Clones have different This", cclne._bshThisCannotClone, clne._bshThisCannotClone);
+ assertNotEquals("Clones have different This namespace", cclne._bshThisCannotClone.namespace, clne._bshThisCannotClone.namespace);
+ assertNotEquals("Clones have different whoami", cclne.whoami(), clne.whoami());
+ assertNotEquals("Clones have different this", cclne.getThis(), clne.getThis());
+
+ assertEquals("Clone's instance equals this", clne, clne.getThis());
+ assertNotEquals("Clone's instance not equal to other clone's this", clne, cclne.getThis());
+ assertThat("Clone's whoami contains instance string", clne.whoami(), containsString(clne.toString()));
+ assertThat("Clone's whoami ends with name string", clne.whoami(), endsWith(clne.name));
+
+ assertEquals("Clones have the same field values", cclne.name, clne.name);
+ clne.name = "iClone";
+ assertNotEquals("Clones can change fields independently", cclne.name, clne.name);
+ assertEquals("Name is what we expect", "iClone", clne.name);
+ assertThat("Name in method output", clne.whoami(), endsWith(clne.name));
+ assertThat("Name still in other method output", cclne.whoami(), endsWith(cclne.name));
+ cclne.name = "oClone";
+ assertNotEquals("Clones can change fields independently", clne.name, cclne.name);
+ assertEquals("Name is what we expect", "oClone", cclne.name);
+ assertThat("Name in method output", cclne.whoami(), endsWith(cclne.name));
+ assertThat("Name still in other method output", clne.whoami(), endsWith(clne.name));
+
+ // cannot create new instance without empty constructor
+
+class NoEmpty {
+    NoEmpty(no) {}
+}
+
+ne = new NoEmpty(1);
+assert(isEvalError("Unable to clone from This reference: NoEmpty", "ne._bshThisNoEmpty.clone();"));
+
+complete();

